models/migrations/v132.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"xorm.io/xorm"
+)
+
+func refixMergeBase(x *xorm.Engine) error {
+	type Repository struct {
+		ID        int64 `xorm:"pk autoincr"`
+		OwnerID   int64 `xorm:"UNIQUE(s) index"`
+		OwnerName string
+		LowerName string `xorm:"UNIQUE(s) INDEX NOT NULL"`
+		Name      string `xorm:"INDEX NOT NULL"`
+	}
+
+	type PullRequest struct {
+		ID         int64 `xorm:"pk autoincr"`
+		Index      int64
+		HeadRepoID int64 `xorm:"INDEX"`
+		BaseRepoID int64 `xorm:"INDEX"`
+		HeadBranch string
+		BaseBranch string
+		MergeBase  string `xorm:"VARCHAR(40)"`
+
+		HasMerged      bool   `xorm:"INDEX"`
+		MergedCommitID string `xorm:"VARCHAR(40)"`
+	}
+
+	var limit = setting.Database.IterateBufferSize
+	if limit <= 0 {
+		limit = 50
+	}
+
+	i := 0
+	for {
+		prs := make([]PullRequest, 0, 50)
+		if err := x.Limit(limit, i).Asc("id").Where("has_merged = ?", true).Find(&prs); err != nil {
+			return fmt.Errorf("Find: %v", err)
+		}
+		if len(prs) == 0 {
+			break
+		}
+
+		i += len(prs)
+		for _, pr := range prs {
+			baseRepo := &Repository{ID: pr.BaseRepoID}
+			has, err := x.Table("repository").Get(baseRepo)
+			if err != nil {
+				return fmt.Errorf("Unable to get base repo %d %v", pr.BaseRepoID, err)
+			}
+			if !has {
+				log.Error("Missing base repo with id %d for PR ID %d", pr.BaseRepoID, pr.ID)
+				continue
+			}
+			userPath := filepath.Join(setting.RepoRootPath, strings.ToLower(baseRepo.OwnerName))
+			repoPath := filepath.Join(userPath, strings.ToLower(baseRepo.Name)+".git")
+
+			gitRefName := fmt.Sprintf("refs/pull/%d/head", pr.Index)
+
+			parentsString, err := git.NewCommand("rev-list", "--parents", "-n", "1", pr.MergedCommitID).RunInDir(repoPath)
+			if err != nil {
+				log.Error("Unable to get parents for merged PR ID %d, Index %d in %s/%s. Error: %v", pr.ID, pr.Index, baseRepo.OwnerName, baseRepo.Name, err)
+				continue
+			}
+			parents := strings.Split(strings.TrimSpace(parentsString), " ")
+			if len(parents) < 3 {
```

Comment:
I think we should consider handling PRs with only one merge base at this point in case a previous version of 128.go would have left some wrong info and one of the bases disappeared in the meantime. But maybe I'm talking nonsense.

Reply:
128 has always been correct if you have a single parent.

Reply:
It's about the parent of the merge commit.

If the commit A has two parents B and C and you ask git to provide the mergebase between A^ i.e. the parent of A and B it will simply reply B, instead of finding the mergebase between B & C.

In the case where there is one parent the calculated mergebase should be correct.

Reply:
I meant that there may have been two parents when 128.go was run (so 128.go could have produced wrong results), but one of the parents is deleted after the fact. But I don't understand this issue very well, so I could be taking it all wrong.

Reply:
Then the merge commit SHA ID would change. These are in git... You can't get rid of the parents without breaking the entire Merkle tree.

Reply:
Oh, I see, this doesn't care about references (i.e. branches, tags, etc.), it only cares about actual commit IDs. Sorry. 😄

Reply:
no prob